### PR TITLE
Subgraph visualisation 🌐 + plotting code re-organisation 📁

### LINKED ISSUE #216: Split & properly orient graph & image visualisation functions
Mainly address this beast: 👇 

https://github.com/alan-turing-institute/grace/blob/58cfa28c79450eb0262cd2c2a40bca3e7ce62ef7/grace/evaluation/visualisation.py#L112-L114

### LINKED ISSUE #227: [DEVELOPMENT] Split the `read_grace.ipynb` viz function
No need to split per se, just use the identical colorbars & invert the y-axis of simple graph plotter. Addresses #216

### LINKED ISSUE #235: [DEVELOPMENT] Separate the `evaluation` and `visualisation` subfolders

### LINKED ISSUE #236: [FEATURE REQUEST] Plot the subgraphs with highlighted nodes & edges per label

### LINKED ISSUE #237: [FEATURE REQUEST] Add function to visualise subgraphs directly from whole graph information

### LINKED ISSUE #266: [DEVELOPMENT] Allow classifier training from `config.yaml` file
- Includes template file
- Parses arguments correctly

### LINKED ISSUE #267: [DEVELOPMENT] Extend `GraphLabelPredictor` to perform inference on entire batch

### LINKED ISSUE #268: [DEVELOPMENT] Return TSNE plot as figure to save / show downstream

### LINKED ISSUE #269: [DEVELOPMENT] Include F1-score into training metrics monitoring in TensorBoard
Accuracy is not reflective of skewed performance towards one (over-dominated) class.

### PR DESCRIPTION
**Pull request by KristinaUlicna**
## PR contribution summary

**Why is this PR useful / good for?** *Please describe the problem(s) you're trying to address.*
* visualise the subgraphs for each node (optionally, with `Prediction` highlighted)
* separates `visualisation` scripts from the `evaluation` subfolder which has grown in the meantime
* extends `GraphLabelPredictor` by methods to calculate inference batch metrics & save/show figures

### List of proposed changes / linked issues & discussions

- [x] Resolves #235 
- [x] Resolves #236 
- [x] Resolves #237 
- [x] Resolves #227
- [x] Resolves #216
- [x] Resolves #266 
- [x] Resolves #267 
- [x] Resolves #268
- [x] Resolves #269

### What should a reviewer concentrate their feedback on?

+ Code quality & correctness, particularly for inference & metrics calculation
+ Notebooks to verify - not just run, please double-check the logic:
    + `read_grace.ipynb` 
    + `infer_predictions.ipynb`
+ Train a model from your accompanying `config.yaml` file

#### What type of PR is this? (check all applicable)

+ 🪄 Feature
+ 🧑‍💻 Code refactor

#### Added tests?

+ 🙅 no, I'll address this in a separate PR

---

## PR review summary

Describe what this PR does & **how you reviewed** the individual items, where needed:

* <TEXT>

*Some helper checks to tick off:*

* Focus on image annotation
* Focus on model training
* Could any optimization be applied?
* Is there any redundant code?
* Are there any spelling errors?

**In conclusion**, after my review, I'd like to:

* 🙋 ask some clarifying questions
* 🙅 suggest some specific changes

### CODE REVIEW

**Comment by KristinaUlicna** [comment]
For more accurate documentation, I'm transferring the description of the config hyperparameters to the `config.yaml` file, which is more flexible & allows insertion of comments & splitting the params into sections 🚀

File: grace/evaluation/inference.py [hunk 1]
```diff
@@ -1,184 +1,264 @@
...
+                data[4].append(e[GraphAttrs.EDGE_PREDICTION].label)
+                data[5].append(e[GraphAttrs.EDGE_PREDICTION].prob_TP)
+
+        # Sanity checks:
+        assert len(data[0]) == len(data[1]) == len(data[2])
+        assert len(data[3]) == len(data[4]) == len(data[5])
+        data = [np.array(item) for item in data]
+
+        n_true, n_pred, n_prob, e_true, e_pred, e_prob = data
+        return n_true, n_pred, n_prob, e_true, e_pred, e_prob
```
Comment (mooniean) [inline grace/evaluation/inference.py:156]: why not return data if in calculate_numerical_results the inverse operation is done?

File: grace/evaluation/inference.py [hunk 2]
```diff
@@ -1,184 +1,264 @@
...
+        return fig
+
+    def get_predictions_for_entire_batch(
+        self,
+        infer_target_list: list[dict[str]],
+        verbose: bool = False,
+    ) -> tuple[npt.NDArray]:
+        """Processes predictions for list of graph targets."""
+
+        data = [[] for _ in range(6)]
```
Comment (mooniean) [inline grace/evaluation/inference.py:126]: nomenclature is a bit confusing as to what "data" is

**Review by mooniean — COMMENTED** [review]
Some initial comments. Will run the code and come back

File: grace/evaluation/inference.py (hunk 2 above)
Comment (KristinaUlicna) [reply grace/evaluation/inference.py:126]: Good point - returning & unwrapping the ordered list of lists is confusing & assumes that the order will always be known. I restructured this into a dictionary where the correct list can be accessed by its string key. This is now fixed in #250: 

https://github.com/alan-turing-institute/grace/blob/a7cf4100766c67b99d167fee8ea63207b1fb41e3/grace/evaluation/inference.py#L156-L170

https://github.com/alan-turing-institute/grace/blob/a7cf4100766c67b99d167fee8ea63207b1fb41e3/grace/evaluation/inference.py#L220-L229

Does this seem more intuitive?

Comment (mooniean) [reply grace/evaluation/inference.py:126]: yes, much better!

**Review by mooniean — APPROVED** [review]
I've ran the notebooks, LGTM.

Just a note: TSNE plot breaks when the model is not good (i.e., nodes are all classified as 0). Issue incoming.

Everything else is working as expected and the plots are really nice!

**Pull request merged**